### PR TITLE
Harden log formatters to handle null formats, null arguments, and null arrays of arguments

### PR DESCRIPTION
**Pull request by natemcmaster**
Resolves https://github.com/aspnet/Logging/issues/422

cc @muratg @BrennanConroy

### CODE REVIEW

File: src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs [hunk 1]
```diff
@@ -114,6 +115,8 @@ private static int FindIndexOf(string format, char ch, int startIndex, int endIn
             return findIndex == -1 ? endIndex : findIndex;
         }
 
+        private static readonly object[] EmptyArray = new object[0];
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs]: Move this to the top of the files with the other fields.

**Comment by pranavkm** [comment]
looks good to me. Have someone more familiar with this code sign off on this though.

**Comment by muratg** [comment]
cc @victorhurdugaci

**Comment by victorhurdugaci** [comment]
:shipit:

File: src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs [hunk 2]
```diff
@@ -20,6 +20,7 @@ public class LogValuesFormatter
 
         public string OriginalFormat { get; private set; }
         public List<string> ValueNames => _valueNames;
+        private const string NullValue = "(null)";
```
Comment (cesarbs) [inline src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs]: nit: move to top of class, or below existing private members.

Comment (cesarbs) [inline src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs]: While you're at it, you could move `OriginalFormat` and `ValueNames` after the ctor

File: test/Microsoft.Extensions.Logging.Test/ConsoleLoggerTest.cs [hunk 3]
```diff
@@ -54,16 +54,27 @@ public ConsoleLoggerTest()
...
+            // Act
+            logger.LogCritical(eventId: 0, exception: null, message: null);
+            logger.LogCritical(eventId: 0, message: null);
+            logger.LogCritical(eventId: 0, message: null, exception: exception);
+
+            // Assert
+            Assert.Equal(expectedMessage, sink.Writes[2].Message);
+            Assert.Equal(expectedMessage, sink.Writes[5].Message);
+            Assert.Equal(expectedMessage, sink.Writes[8].Message);
+            Assert.Equal(expectedExceptionMessage, sink.Writes[9].Message);
```
Comment (cesarbs) [inline test/Microsoft.Extensions.Logging.Test/ConsoleLoggerTest.cs]: Magic numbers 😁  Isn't there a less-potentially-flaky way to capture those messages?

File: src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs (hunk 1 above)
Comment (moozzyk) [inline src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs]: Why not just `Array.Empty<object>()` ?

Comment (davidfowl) [inline src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs]: low netstandard version

Comment (moozzyk) [inline src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs]: I see.

Comment (natemcmaster) [inline src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs]: btw, is the `s_` variable name prefix for static/const privates `s_varName` a convention we do on this team?

**Pull request merged**